src/qe/mbp/mbp_arrays_tg.cpp

```diff
@@ -0,0 +1,367 @@
+                expr_ref npeq(mk_not(p_new.mk_peq()), m);
+                m_tg.add_lit(npeq);
+                m_tg.add_eq(p.mk_peq(), p_new.mk_peq());
+            }
+            if (!m_mdl.are_equal(rd, elem)) {
+                m_tg.add_deq(rd, elem);
+            }
+        }
+    }
+
+    void add_rdVar(expr* rd) {
+        //do not assign new variable if rd is already equal to a value
+        if (m_tg.has_val_in_class(rd)) return;
+        TRACE("mbp_tg", tout << "applying add_rdVar on " << expr_ref(rd, m););
+        app_ref u = new_var(to_app(rd)->get_sort(), m);
+        m_new_vars.push_back(u);
+        m_tg.add_var(u);
+        m_tg.add_eq(u, rd);
+        m_mdl.register_decl(u->get_decl(), m_mdl(rd));
+    }
+
+    void elimeq(peq p) {
+        TRACE("mbp_tg", tout << "applying elimeq on " << expr_ref(p.mk_peq(), m););
+        app_ref_vector aux_consts(m);
+        expr_ref eq(m);
+        expr_ref sel(m);
+        eq = p.mk_eq(aux_consts, true);
+        vector<expr_ref_vector> indices;
+        p.get_diff_indices(indices);
+        vector<expr_ref_vector>::iterator itr = indices.begin();
+        unsigned i = 0;
+        for(app* a : aux_consts) {
+            m_new_vars.push_back(a);
+            m_tg.add_var(a);
+            auto const& indx =  std::next(itr, i);
+            SASSERT(indx->size() == 1);
+            expr *args[2] = {to_app(p.lhs()), to_app(indx->get(0))};
+            sel = m_array_util.mk_select(2, args);
+            m_mdl.register_decl(a->get_decl(), m_mdl(sel));
+            i++;
+        }
+        m_tg.add_lit(eq);
+        m_tg.add_lit(m.mk_true());
+        m_tg.add_eq(p.mk_peq(), m.mk_true());
+        TRACE("mbp_tg", tout << "added lit  " << eq;);
+    }
+
+    void elimrdwr(expr* term) {
+        SASSERT(is_rd_wr(term));
+        TRACE("mbp_tg", tout << "applying elimrdwr on " << expr_ref(term, m););
+        expr* wr_ind = to_app(to_app(term)->get_arg(0))->get_arg(1);
+        expr* rd_ind = to_app(term)->get_arg(1);
+        expr* e;
+        if (m_mdl.are_equal(wr_ind, rd_ind)) {
+            m_tg.add_eq(wr_ind, rd_ind);
+            e =  to_app(to_app(term)->get_arg(0))->get_arg(2);
+        }
+        else {
+            m_tg.add_deq(wr_ind, rd_ind);
+            expr* args[2] = {to_app(to_app(term)->get_arg(0))->get_arg(0), to_app(term)->get_arg(1)};
+            e = m_array_util.mk_select(2, args);
+        }
+        m_tg.add_eq(term, e);
+    }
+
+// iterate through all terms in m_tg and apply all array MBP rules once
+// returns true if any rules were applied
+    bool apply() {
+        TRACE("mbp_tg", tout << "Iterating over terms of tg";);
+        indices.reset();
+        rdTerms.reset();
+        m_new_vars.reset();
+        expr_ref e(m), rdEq(m), rdDeq(m);
+        expr* nt, *term;
+        bool progress = false, is_neg = false;
+
+        //Not resetting terms because get_terms calls resize on terms
+        m_tg.get_terms(terms, false);
```

Comment:
why do you need to copy all expression references into a new vector, can you return an iterator?

Reply:
The term graph only stores terms, not expressions of terms. This is my attempt at not exposing terms